### ISSUE TITLE
Ray exceptions should be more specific

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#4236]
https://github.com/ray-project/ray/blob/e2e6ef198bb281151b7fd0ab478d44b2066c8f55/python/ray/experimental/state.py#L108-L121

Here, the exception is so general (i.e., what if something else failed?)

Ray should have specific exception classes for applications to catch.

**Comment by zbarry** [comment]
See #4090 :). Was going to add a PR to update the exception in `worker.py` here:

```python
    relevant_client = None
    for client_info in client_table:
        client_node_ip_address = client_info["NodeManagerAddress"]
        if (client_node_ip_address == node_ip_address or
            (client_node_ip_address == "127.0.0.1"
             and redis_ip_address == ray.services.get_node_ip_address())):
            relevant_client = client_info
            break
    if relevant_client is None:
        raise Exception(
            "Redis has started but no raylets have registered yet.")
```

To account for (as I understand it) what it really means is there is no Ray worker / head on the machine that you are `ray.init()`ing from. I was confused for the longest time thinking that for whatever reason worker nodes I've spawned with the autoscaler weren't being registered properly with the head when my driver's init failed.

**Comment by richardliaw** [comment]
Addressed in #7523
